### PR TITLE
v2 permalink compatibility for background layers, layers and opacity

### LINKED ISSUE #555: Permalink backwards compatibility

### PR DESCRIPTION
**Pull request by petzlux**
fix #555

### CODE REVIEW

**Comment by petzlux** [comment]
fixes #555

**Comment by petzlux** [comment]
Ready for review

**Comment by petzlux** [comment]
@rmichaelis can you review ?

**Comment by rmichaelis** [comment]
V2 permalink does not contain the bgLayer parameter when the background is not modified. 
Because the default basemap is not the same in both V2 and V3, the result is different. 
It's a difference but probably not a problem.

**Comment by jaykayone** [comment]
good remark, @rmichaelis 

None should translate to the default of the theme, which can be found in the metadata of the theme

**Comment by petzlux** [comment]
@jaykayone I dont think that the theme bg_layer property is taken into account at the moment?

**Comment by petzlux** [comment]
@rmichaelis good to merge otherwise ? we will address background layer in separate pull request, see #635

**Comment by rmichaelis** [comment]
why are these 2 maps so different ? 
http://devv3rm.geoportail.lu/?zoom=4&X=75098.7415&Y=76009.28129&layers=addresses&layers_indices=6&layers_opacity=1&layers_visibility=true&bgLayer=webbasemap

http://map.geoportail.lu/?zoom=4&X=75098.7415&Y=76009.28129&layers=addresses&layers_indices=6&layers_opacity=1&layers_visibility=true&bgLayer=webbasemap

**Comment by jaykayone** [comment]
because the zoomlevel lookup seems to be one off

**Comment by petzlux** [comment]
No its more subtle than that ... the zoomlevels are not equivalent ...

**Comment by jaykayone** [comment]
yes that's for sure but in the particular case, the smaller one would be a much better match

**Comment by petzlux** [comment]
i will try and find a compromise

**Comment by rmichaelis** [comment]
but in this case, scales are the same ?

**Comment by petzlux** [comment]
@rmichaelis @jaykayone can you check if the new zoom level conversion is okay

**Comment by jaykayone** [comment]
bgOpacity=0 should switch to ortho background

**Comment by jaykayone** [comment]
zoomlevels are perfect.

in some thematic portals (like go), there are additional levels (11+). we should at least have a fallback to 19 for those

**Comment by petzlux** [comment]
@jaykayone added your feature requests.

**Comment by petzlux** [comment]
Ready for review

**Comment by jaykayone** [comment]
works fine

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 1]
```diff
@@ -270,8 +288,71 @@ app.LayerPermalinkManager.prototype.init =
 
         this.appThemes_.getThemesObject().then(
             goog.bind(function(themes) {
-              if (goog.isDef(layerIds)) {
-                this.applyLayerStateToMap_(layerIds, themes);
+              var flatCatalogue = [];
+              for (var i = 0; i < themes.length; i++) {
+                var theme = themes[i];
+                goog.array.extend(flatCatalogue,
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: `goog.require('goog.array')` missing.

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 2]
```diff
@@ -270,8 +288,71 @@ app.LayerPermalinkManager.prototype.init =
...
+              var layerIds = [];
+              /**
+               * @type {Array.<number>|undefined}
+               */
+              var opacities = [];
+
+              if (this.initialVersion_ === 2) {
+                var layerString = this.stateManager_.getInitialValue('layers');
+                if (goog.isDefAndNotNull(layerString) &&
+                    !goog.string.isEmpty(layerString)) {
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: `goog.require('goog.string')` missing.

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 1 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: already there

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 2 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: will add in new pull request

**Pull request merged**